### ISSUE TITLE
[SPIR-V] Handle arbitrarily deep wrapping of `StructuredBuffer`s

### ISSUE THREAD

**Issue by sudonatalie** [microsoft/DirectXShaderCompiler#4187]
Rewrite `RemoveBufferBlockVisitor` to handle arbitrarily deep nesting of `StructuredBuffer`s inside structs/objects.

Following the fix for #4185, the visitor will be able to address:
```
  /// 1- a pointer to a structure with StorageBuffer interface
  /// 2- a pointer to a pointer to a structure with StorageBuffer interface
  /// 3- a pointer to a struct containing a structure with StorageBuffer interface
```
but the logic should be re-implemented to handle arbitrary nesting of structs/pointers. This logic may be better suited to a new legalization pass if it can be completed bottom-up more efficiently.

**Comment by sudonatalie** [comment]
@jaebaek From a DXC design perspective, does it make sense for this visitor to simply be rewritten to traverse variable and parameter types all the way down, or does logic like that generally get moved to a legalization pass? I'm not sure if there's a reason it was written this way in the first place (only handling pointers one level down).

**Comment by jaebaek** [comment]
This decision is depending on the situation.
If the code transformation needs HLSL information that is not included in SPIR-V, it must be in DXC. We usually use multiple "visitors" after the AST traversal.

> From a DXC design perspective, does it make sense for this visitor to simply be rewritten to traverse variable and parameter types all the way down, or does logic like that generally get moved to a legalization pass?

The former is fine. In this case, the legalization must not handle all the cases because there can be an actual error case e.g., initial HLSL code was wrong or other high-level language like Jaebaek-LSL can have initially wrong code gen.

Since the legalization will remove the load/store of resource types only when the temporary variable has the same storage class, DXC must generate the code whose load/store resource types match each other.

**Comment by jaebaek** [comment]
My concern on `RemoveBufferBlockVisitor` is that it handles only a few cases. It must handle other cases e.g., recursive array/struct including resources.

**Comment by jaebaek** [comment]
I agree with you. Moving it to a spirv-opt pass with be better for maintenance in the long term because spirv-opt has the good infrastructure including the type analysis.

However, I do not think it should be a part of the legalization.
It must be just a separate spirv-opt pass and DXC must run the pass after SPIR-V code gen when the Vulkan version is 1.2 or above.

If you ask me the initial design decision, I was not the author and .. probably we did not know the legalization is sensitive about the storage class. The `RemoveBufferBlockVisitor` was a part of VK1.2 support. I believe we only focus on VK1.2, but we did not imagine the side effect caused by the storage class mismatch.
